### PR TITLE
feat: ORT GenAI Stateful Compilation changes

### PR DESCRIPTION
**Pull request by ankitm3k**
### Description
This PR enables  the essential features to enable ORT GenAI with OVEP using Stateful Compilation of ov::Model, inspired from OV GenAI pipeline flow.

I have introduced a new provider option `enable_causallm` which can be set to `True` for enabling the ORT GenAI pipeline with Causal LLM _Models that are fully supported on OVEP_ in the custom config file called _genai_config.json_ 

Sample _genai_config.json_ -

```
"provider_options": [
    {
         // Key "OpenVINO" is case sensitive and must be used as below as its defined by MSFT
         "OpenVINO":
          {
              "device_type": "NPU",
              "enable_causallm" : "True",
              // (Applicable for NPU only) Optional setting for compilation with custom MAX_PROMPT_LEN & MIN_RESPONSE_LEN
              "load_config": "{\"NPU\":{\"MAX_PROMPT_LEN\":\"2048\",\"MIN_RESPONSE_LEN\":\"512\"}}"                            
                       
           }
    }
    ]
```          
FYI the GenAI models in ONNX format are usually Stateless in nature & require dynamic shapes compilation.

### CODE REVIEW

File: onnxruntime/core/providers/openvino/backends/basic_backend.cc [hunk 1]
```diff
@@ -374,14 +393,22 @@ void BasicBackend::StartAsyncInference(Ort::KernelContext& context, OVInferReque
         }
         index++;
       }
+
+      // For Stateful Model Compilation, the ONNX model includes KV cache (past/present) tensors.
+      // However, these tensors are internally converted to a stateful representation, which removes them.
+      // To prevent runtime exceptions, we simply continue processing here.
+      if (input_name.empty() || input_name == "beam_idx") continue;
```
Comment (preetha-intel) [inline onnxruntime/core/providers/openvino/backends/basic_backend.cc:400]: Is the name "beam_idx" particular to LLM model? It looks like a custom check

Comment (ankitm3k) [reply onnxruntime/core/providers/openvino/backends/basic_backend.cc:400]: yes its a input name which is a part of ov::model which gets created as a part of translation. It required for OV LLMCOmpilation & inference pipeline

File: onnxruntime/core/providers/openvino/backend_manager.cc [hunk 2]
```diff
@@ -567,5 +570,11 @@ void BackendManager::ShutdownBackendManager() {
   concrete_backend_.reset();
 }
 
+void BackendManager::RewindKVCache(size_t index) {
+  if (concrete_backend_) {
+    concrete_backend_->RewindKVCache(index);
+  }
+}
```
Comment (preetha-intel) [inline onnxruntime/core/providers/openvino/backend_manager.cc:577]: What if concrete backend doesn't exist? should there be an exception

File: onnxruntime/core/providers/openvino/backend_manager.cc [hunk 3]
```diff
@@ -106,7 +106,8 @@ BackendManager::BackendManager(SessionContext& session_context,
     subgraph_context_.has_dynamic_input_shape = true;
     LOGS_DEFAULT(INFO) << "[OpenVINO-EP] Model has symbolic input dims";
     if ((session_context_.device_type.find("CPU") != std::string::npos ||
-         session_context_.device_type.find("GPU") != std::string::npos) &&
+         session_context_.device_type.find("GPU") != std::string::npos ||
+         (session_context_.device_type.find("NPU") != std::string::npos && session_context_.enable_causallm)) &&
```
[the remark this answers is not in the document]
Comment (ankitm3k) [reply onnxruntime/core/providers/openvino/backend_manager.cc]: fixed

File: onnxruntime/core/providers/openvino/backend_manager.cc (hunk 2 above)
Comment (ankitm3k) [reply onnxruntime/core/providers/openvino/backend_manager.cc:579]: Since GetOVCompiledModel() already handles nullptr by returning an empty model, it would be more consistent to make RewindKVCache a no-operation when the backend doesn't exist. Thus if concrete_backend_ is nullptr then it would do nothing in this case.

File: onnxruntime/core/providers/openvino/ov_interface.cc [hunk 4]
```diff
@@ -277,35 +352,164 @@ void OVInferRequest::StartAsync() {
...
+    {
+      auto position_ids = ovInfReq.get_tensor("position_ids");
+      auto* pData = position_ids.data<int64_t>();
+      for (size_t i = 0; i < position_ids.get_size(); i++) {
+        cached_position_ids.push_back(pData[i]);
+      }
+    }
+
+    // If we're about to run the prefill model
+    if (input_ids_tensor.get_size() > 1) {
```
[the remark this answers is not in the document]
Comment (ankitm3k) [reply onnxruntime/core/providers/openvino/ov_interface.cc]: fixed

File: onnxruntime/core/providers/openvino/ov_interface.cc [hunk 5]
```diff
@@ -277,35 +352,190 @@ void OVInferRequest::StartAsync() {
...
+    prefill_use_full_chat_history = true;
+  }
+}
+
+void StatefulOVInferRequest::CacheTensor(const std::string& tensor_name, const ov::element::Type& type,
+                                         const std::vector<size_t>& shape, int32_t fill_value) {
+  ov::Tensor tensor = ov::Tensor(type, shape);
+  std::fill_n(tensor.data<int32_t>(), tensor.get_size(), fill_value);
+  ovInfReq.set_tensor(tensor_name, tensor);
+}
```
Comment (RyanMetcalfeInt8) [inline onnxruntime/core/providers/openvino/ov_interface.cc:399]: Should this be called 'FillTensor'? It doesn't seem to *cache* anything, right?

File: onnxruntime/core/providers/openvino/ov_interface.cc [hunk 6]
```diff
@@ -277,35 +352,190 @@ void OVInferRequest::StartAsync() {
...
+
+void StatefulOVInferRequest::CacheTensor(const std::string& tensor_name, std::vector<int64_t>& cache) {
```
Comment (RyanMetcalfeInt8) [inline onnxruntime/core/providers/openvino/ov_interface.cc:401]: Note that we may eventually need to support caching position_ids / logits which have more complicated shapes than just `[1, <num_input_tokens>]`.

File: onnxruntime/core/providers/openvino/backends/basic_backend.h [hunk 7]
```diff
@@ -48,6 +48,15 @@ struct OnnxToOvNetworkBindings {
         auto it = std::find_if(ov_parameters.begin(), ov_parameters.end(),
                                [&onnx_name](const auto& ov_parameter_info) { return ov_parameter_info.get_names().contains(onnx_name); });
 
+        // For Stateful Model Compilation, the ONNX model includes KV cache (past/present) tensors.
+        // However, these tensors are internally converted to a stateful representation, which removes them.
+        // To prevent runtime exceptions, we simply continue processing here.
+        if (onnx_name.empty() || onnx_name == "beam_idx" ||
+            onnx_name.find("past_key_values") != std::string::npos ||
+            onnx_name.find("present") != std::string::npos) {
+          continue;
```
Comment (RyanMetcalfeInt8) [inline onnxruntime/core/providers/openvino/backends/basic_backend.h:57]: Should there be an additional `&& enable_causallm` here, so that we only skip this when the stateful flow is enabled?

File: onnxruntime/core/providers/openvino/ov_interface.cc (hunk 5 above)
Comment (ankitm3k) [reply onnxruntime/core/providers/openvino/ov_interface.cc:399]: fixed

File: onnxruntime/core/providers/openvino/backends/basic_backend.h (hunk 7 above)
Comment (ankitm3k) [reply onnxruntime/core/providers/openvino/backends/basic_backend.h:58]: fixed

File: onnxruntime/core/providers/openvino/ov_interface.cc (hunk 6 above)
Comment (ankitm3k) [reply onnxruntime/core/providers/openvino/ov_interface.cc:373]: will handle that logic in the next PR
